lib/vrl/compiler/src/vm/state.rs

```diff
@@ -0,0 +1,83 @@
+use super::{argument_list::VmArgument, machine::Instruction, OpCode, Vm};
+use crate::{ExpressionError, Value};
+
+/// `VmState` contains the mutable state used to run the Vm.
+pub struct VmState<'a> {
+    vm: &'a Vm,
+    /// The index of the next instruction to run.
+    pub(super) instruction_pointer: usize,
+    /// The stack used to store values whilst running the process.
+    pub(super) stack: Vec<Value>,
+    /// A stack of values passed to functions when called.
+    pub(super) parameter_stack: Vec<Option<VmArgument<'a>>>,
+    /// Errors generated by the last expression are stored here.
+    pub(super) error: Option<ExpressionError>,
+}
+
+impl<'a> VmState<'a> {
+    pub(super) fn new(vm: &'a Vm) -> Self {
+        Self {
+            vm,
+            instruction_pointer: 0,
+            stack: Vec::new(),
+            parameter_stack: Vec::new(),
+            error: None,
+        }
+    }
+
+    /// Return the Opcode at the current position and advance the instruction pointer.
+    /// Returns an error if the current position is not an OpCode.
+    /// This error should never occur for a correctly compiled program!
+    pub(super) fn next_opcode(&mut self) -> Result<OpCode, ExpressionError> {
+        let byte = self.vm.instructions[self.instruction_pointer];
+        self.instruction_pointer += 1;
+        match byte {
+            Instruction::OpCode(opcode) => Ok(opcode),
+            _ => Err(format!("Expecting opcode at {}", self.instruction_pointer - 1).into()),
+        }
+    }
+
+    /// Return the primitive value at the current position and advance the instruction pointer.
+    /// Returns an error if the current position is not a primitve value.
+    /// This error should never occur for a correctly compiled program!
+    pub(super) fn next_primitive(&mut self) -> Result<usize, ExpressionError> {
+        let byte = self.vm.instructions[self.instruction_pointer];
+        self.instruction_pointer += 1;
+        match byte {
+            Instruction::Primitive(primitive) => Ok(primitive),
+            _ => Err(format!("Expecting primitive at {}", self.instruction_pointer - 1).into()),
+        }
+    }
+
+    /// Pushes the given value onto the stack.
+    pub(super) fn push_stack(&mut self, value: Value) {
+        self.stack.push(value);
+    }
+
+    /// Pops the value from the top of the stack.
+    /// Errors if the stack is empty.
+    pub(super) fn pop_stack(&mut self) -> Result<Value, ExpressionError> {
+        self.stack.pop().ok_or_else(|| "stack underflow".into())
```

Comment:
better error messages to users + easier to assert against in tests probably

Reply:
That's true, however ideally this is not an error the user should ever see. The user did nothing wrong, this is a programmer error in code generation and should have no cost at normal operation.

I guess in practice the cost of `unwrap` and `Result` in this particular case should be the same though, since I don't think Rust can elide handling that error even in release builds.

Reply:
> better error messages to users + easier to assert against in tests probably

Yeah, exactly this.

Especially when running against VRL tests on an error it just panics and doesn't say what test failed which makes devving frustrating. An optimisation once we are happy with the state of things could be to `unwrap` and see what effect it has on performance..